src/utils/task_stats_alloc/src/lib.rs

```diff
@@ -0,0 +1,202 @@
+    #[inline(always)]
+    unsafe fn add_unchecked(&self, val: usize) {
+        self.0.unwrap_unchecked().fetch_add(val, Ordering::Relaxed);
+    }
+
+    /// Subtracts from the counter value, and `drop` the counter while the count reaches zero.
+    #[inline(always)]
+    fn sub(&self, val: usize) {
+        if let Some(bytes) = self.0 {
+            // Use Release to synchronize with the below deletion.
+            let old_bytes = bytes.fetch_sub(val, Ordering::Release);
+            // If the counter reaches zero, delete the counter. Note that we've ensured there's no
+            // zero deltas in `wrap_layout`, so there'll be no more uses of the counter.
+            if old_bytes == val {
+                // This fence is needed to prevent reordering of use of the counter and deletion of
+                // the counter. Because it is marked `Release`, the decreasing of the counter
+                // synchronizes with this `Acquire` fence. This means that use of the counter
+                // happens before decreasing the counter, which happens before this fence, which
+                // happens before the deletion of the counter.
+                fence(Ordering::Acquire);
+                unsafe { Box::from_raw_in(bytes.as_mut_ptr(), System) };
+            }
+        }
+    }
+
+    #[inline(always)]
+    pub fn val(&self) -> usize {
+        self.0
+            .as_ref()
+            .expect("bytes is invalid")
+            .load(Ordering::Relaxed)
+    }
+}
+
+task_local! {
+    pub static BYTES_ALLOCATED: TaskLocalBytesAllocated;
+}
+
+pub async fn allocation_stat<Fut, T, F>(future: Fut, interval: Duration, mut report: F) -> T
+where
+    Fut: Future<Output = T>,
+    F: FnMut(usize),
+{
+    BYTES_ALLOCATED
+        .scope(TaskLocalBytesAllocated::new(), async move {
+            // The guard has the same lifetime as the counter so that the counter will keep positive
+            // in the whole scope. When the scope exits, the guard is released, so the counter can
+            // reach zero eventually and then `drop` itself.
+            let _guard = Box::new(114514);
+            let monitor = async move {
+                let mut interval = tokio::time::interval(interval);
+                loop {
+                    interval.tick().await;
+                    BYTES_ALLOCATED.with(|bytes| report(bytes.val()));
+                }
+            };
+            let output = tokio::select! {
+                biased;
+                _ = monitor => unreachable!(),
+                output = future => output,
+            };
+            output
+        })
+        .await
+}
+
+#[inline(always)]
+fn wrap_layout(layout: Layout) -> (Layout, usize) {
+    debug_assert_ne!(layout.size(), 0, "the size of layout must be non-zero");
+
+    let (wrapped_layout, offset) = Layout::new::<TaskLocalBytesAllocated>()
+        .extend(layout)
+        .expect("wrapping layout overflow");
+    let wrapped_layout = wrapped_layout.pad_to_align();
```

Comment:
Add a check here to ensure new alignment matches old alignment? cc @wangrunji0408 @BowenXiao1999

Reply:
Seems that `extend` method takes it into account.

Reply:
What exact variables of 'old' and 'new' do you mean?

If you mean this line:
```rust
let wrapped_layout = wrapped_layout.pad_to_align();
//  ^ new            ^ old
```
The alignment won't be changed by definition and by its implementation.

If you mean the function:
```rust
fn wrap_layout(layout: Layout) -> (Layout, usize)
//             ^ old               ^ new
```
I guess it is possible to be changed?

Reply:
The later one. But according to comments in `extend` methods, I think old layout's alignment is still respected.

Reply:
But if the old layout comes from a `u8` (size = 1, align = 1), the new layout should be `(usize, u8)` (size = 16, align = 8)?